specs/chugsplash-config.md

```diff
@@ -0,0 +1,359 @@
+
+Note that we explicitly do not support alternative base inputs such as hexadecimal for uint and int.
+
+### Examples
+```js
+// Solidity
+uint public x;
+
+// Javascript
+x: "1",
+```
+
+```js
+// Solidity
+uint public x;
+
+// Javascript
+x: 1,
+```
+
+```js
+// Solidity
+int public x;
+
+// Javascript
+x: -1,
+```
+
+```js
+// Solidity
+int public x;
+
+// Javascript
+x: "-1",
+```
+
+## Booleans
+For booleans, we only allow literal Javascript or JSON booleans.
+
+### Examples
+```js
+// Solidity
+boolean public x;
+boolean public y;
+
+// Javascript
+x: true,
+y: false,
+```
+
+## Addresses
+For addresses, we only accept hexadecimal literals that pass the address checksum test.
+
+### Example
+```js
+// Solidity
+address public x;
+
+// Javascript
+x: '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266',
+```
+
+## Contracts
+For contracts, we accept two possible input types:
+- Valid checksum addresses as defined above
+- ChugSplash contract references
+
+ChugSplash contract references are a format that allows users to reference other contracts deployed using ChugSplash. Our offchain tooling converts these references into the underlying contract address using Create2.
+
+Valid contract references must reference contracts in the same ChugSplash project and use the following format:
+```
+{{ <Contract Reference Name> }}
+```
+
+### Example
+```js
+// Solidity
+Contract public contractReference;
+
+// Javascript
+const config: UserChugSplashConfig = {
+  // Configuration options for the project:
+  options: {
+    projectName: 'My Project',
+  },
+  contracts: {
+    Contract: {
+      contract: 'Contract',
+      ...
+    },
+    OtherContract: {
+      contract: 'SomeOtherContract',
+      variables: {
+        contractReference: '{{ Contract }}',
+      },
+    },
+  },
+}
+```
+
+## Strings
+For strings, we allow only unicode string literals.
+Note that if you supply a hex value in string format, we will not interpret it as a bytes value. We will encode it as if it were any other string.
+
+### Example
+```js
+// Solidity
+string public x;
+
+// Javascript
+x: 'literalString',
+```
+
+## User Defined Types
+We treat user defined types as if they were their underlying value type. We expect the input format to match the expected format of the underlying value type.
+
+### Example
+```js
+// Solidity
+type UserDefinedType is uint256;
+UserDefinedType public x;
+
+// Javascript
+x: 1,
+```
+
+## Enums
+Similar to user defined types, we treat enums like their underlying value type uint8. We accept the following possible input formats:
+- Base 10 integers between 0-255 where the integer corresponds to a valid Solidity enum value.
```

Comment:
Instead of 0-255, it should actually be something like "must be an integer N for which:
```
0 <= N < type(NameOfEnum).max
```
In other words, if there are three members in an enum, our validation logic should throw an error if the user enters `255`.